### PR TITLE
Array string distance alpha

### LINKED ISSUE #1994: [FEAT] Allow fuzzy matches on array-valued columns
### Is your proposal related to a problem?

One typical use of array-valued columns is when one known entity has multiple values for a single attribute, a match on any of which is acceptable. For example one person may have multiple telephone numbers or an address history, or even multiple names (married/unmarried names, aliases, etc). While exploding the records outside of splink and post-processing matches is one alternative to array-valued columns, this affects match probabilities and may create memory challenges. 

Currently the only way to compare array-valued columns is to use the array_intersect functions, which require exact matches among the elements of the array.

However, in many circumstances, these values are prone to error and fuzzy matching may be required to identify matches.

### Describe the solution you'd like

I propose a family of matching functions of the form array_min_sim(x, y, simtype) and array_max_sim(x,y,simtype). Here x and y are arrays, and "simtype" is a string identifying one of the existing fuzzy string comparison metrics (alternatively, a separate array function could be implemented for each metric). 

The semantics would be to return the maximum or minimum similarity (alternatively, distance) between an element of array x and an element of array y.  These max/min similarities could be thresholded to get comparison levels.

I'm not sure what the best implementation of this would be. Duckdb may allow an implementation using an iterated call to list_reduce. For example, for array_min_sim, the underlying code could be a list_reduce applied to [Infinity,x] (Infinity prepended to x), with reduce function (a, b) -> min(a, list_reduce( [Infinity,y], (c,d) -> min(c, dist(b,d ) ) .

Alternatively, the arrays could be exploded locally and a full cartesian product could be constructed temporarily (with no other variables to reduce memory overhead), the string comparison run on on the expanded data, then aggregated with min/max; I believe this is similar to what is/was being considered for array valued blocking keys (#1448) (though I can't say I understand the internals enough to understand if this will work!)

### Describe alternatives you've considered

Un-nesting the data before running splink and post-processing is one option, but this affects m-values and creates memory challenges via duplicated records. 

Using array intersections as another alternative, but it demands exact matching.

### PR DESCRIPTION
**Pull request by JonnyShiUW**
… distance feature

### Type of PR

- [ ] BUG
- [x] FEAT
- [ ] MAINT
- [ ] DOC

### Is your Pull Request linked to an existing Issue or Pull Request?
  Closes #1994 

### Give a brief description for the solution you have provided

First prototype of fuzzy matching for array-value columns for DuckDB

### PR Checklist

- [ ] Added documentation for changes
- [ ] Added feature to example notebooks or tutorial (if appropriate)
- [ ] Added tests (if appropriate)
- [ ] Updated CHANGELOG.md (if appropriate)
- [x] Made changes based off the latest version of Splink
- [ ] Run the [linter](https://moj-analytical-services.github.io/splink/dev_guides/changing_splink/lint_and_format.html)
- [ ] Run the [spellchecker](https://moj-analytical-services.github.io/splink/dev_guides/changing_splink/contributing_to_docs.html?h=spellch#spellchecking-docs) (if appropriate)

### CODE REVIEW

**Comment by zmbc** [comment]
Splink folks: this is a work in progress! I'll give it a review first.

File: splink/internals/comparison_level_library.py [hunk 1]
```diff
@@ -742,6 +742,66 @@ def create_sql(self, sql_dialect: SplinkDialect) -> str:
...
+            parameter_value=distance_threshold,
+            level_name=self.__class__.__name__,
+            parameter_name="distance_threshold",
+        )
+        self.distance_function = validate_categorical_parameter(
+            allowed_values=["levenshtein", "damerau_levenshtein", "jaro_winkler", "jaro"],
+            parameter_value=distance_function,
+            level_name=self.__class__.__name__,
+            parameter_name="distance_function"
+        )
```
Comment (zmbc) [inline splink/internals/comparison_level_library.py:772]: I think it's an open question whether it is better for the user to define any function name they want (as in `DistanceFunctionLevel`) or only to have certain options (but it gets auto-transpiled).

File: splink/internals/comparison_level_library.py [hunk 2]
```diff
@@ -742,6 +742,66 @@ def create_sql(self, sql_dialect: SplinkDialect) -> str:
...
+                                {col.name_l}, 
+                                x -> list_transform(
+                                    {col.name_r}, 
+                                    y -> [x,y]
+                                )
+                            )
+                        ), 
+                        pair -> {d_fn}(pair[1], pair[2])
+                    )
+                ) <= {self.distance_threshold}"""
```
Comment (zmbc) [inline splink/internals/comparison_level_library.py:800]: This comparison should likely change to a `>=` for "jaro" and "jaro_winkler" where higher scores are more similar.

File: splink/internals/comparison_library.py [hunk 3]
```diff
@@ -535,6 +535,39 @@ def datetime_parse_function(self):
     def cll_class(self):
         return cll.AbsoluteDateDifferenceLevel
 
+class ArrayStringDistance(ComparisonCreator):
+    def __init__(
+            self,
+            col_name: str,
+            distance_threshold_or_thresholds: Union[Iterable[int], int] = [1],
+            distance_function: str = "levenshtein",
```
Comment (zmbc) [inline splink/internals/comparison_library.py:543]: This should probably be a custom type that lists the allowable values, similar to `DateMetricType` above.

File: splink/internals/comparison_level_library.py [hunk 4]
```diff
@@ -742,6 +742,66 @@ def create_sql(self, sql_dialect: SplinkDialect) -> str:
...
+        )
+    
+    def create_label_for_charts(self) -> str:
+        return f"Array string distance <= {self.distance_threshold}"
```
Comment (zmbc) [inline splink/internals/comparison_level_library.py:804]: Let's include the distance function name here

File: splink/internals/comparison_library.py [hunk 5]
```diff
@@ -535,6 +535,39 @@ def datetime_parse_function(self):
...
+                for threshold in self.thresholds
+            ],
+            cll.ElseLevel(),
+        ]
+
+    def create_description(self) -> str:
+        comma_separated_thresholds_string = ", ".join(map(str, self.thresholds))
+        plural = "s" if len(self.thresholds) > 1 else ""
+        return (
+            f"Array string distance at maximum size{plural} "
```
Comment (zmbc) [inline splink/internals/comparison_library.py:565]: Let's include the distance function name here

File: splink/internals/comparison_level_library.py (hunk 2 above)
Comment (zmbc) [reply splink/internals/comparison_level_library.py:800]: Oh and distance_threshold should probably be a float, not an int.

File: splink/internals/comparison_level_library.py [hunk 6]
```diff
@@ -742,6 +742,66 @@ def create_sql(self, sql_dialect: SplinkDialect) -> str:
...
+            d_fn = sql_dialect.jaro_function_name
+        return (
+            f"""list_max(
+                    list_transform(
+                        flatten(
+                            list_transform(
+                                {col.name_l}, 
+                                x -> list_transform(
+                                    {col.name_r}, 
+                                    y -> [x,y]
```
Comment (zmbc) [inline splink/internals/comparison_level_library.py:794]:
```suggestion
                                l_item -> list_transform(
                                    {col.name_r}, 
                                    r_item -> [l_item, r_item]
```

Just for readability

**Review by zmbc — COMMENTED** [review]
Thanks @JonnyShiUW! Looks good overall, with a few minor things to work out. We also need to add some tests.

**Comment by RobinL** [comment]
Thanks!  I'm on leave after today, but i had a very quick look, and tried it out:

<details>
<summary>
Runnable code to try new comparison
</summary>

```python
import duckdb
import pandas as pd

import splink.internals.comparison_level_library as cl
from splink import DuckDBAPI, Linker, SettingsCreator
from splink.internals.comparison_library import CustomComparison
from splink.internals.dialects import DuckDBDialect

data_l = pd.DataFrame.from_dict(
    [
        {"unique_id": 1, "name": ["robin", "james"]},
        {"unique_id": 2, "name": ["robyn", "steve"]},
        {"unique_id": 3, "name": ["stephen"]},
        {"unique_id": 4, "name": ["stephen"]},
    ]
)

arr_comparison = cl.ArrayStringDistanceLevel("name", 2, "levenshtein")
cc = CustomComparison(
    [
        cl.NullLevel("name"),
        cl.ExactMatchLevel("name"),
        arr_comparison,
        cl.ElseLevel(),
    ]
)
print(arr_comparison.create_sql(DuckDBDialect()))

settings = SettingsCreator(
    link_type="dedupe_only",
    blocking_rules_to_generate_predictions=["1=1"],
    comparisons=[cc],
)


linker = Linker(data_l, settings, database_api=DuckDBAPI(), set_up_basic_logging=True)


linker.predict().as_pandas_dataframe()


# Test sql
literal_array_l = str(["robin", "james"])
literal_array_r = str(["robyn", "bob"])


sql = f"""
select
    list_max(
        list_transform(
            flatten(
                list_transform(
                    {literal_array_l},
                    x -> list_transform(
                        {literal_array_r},
                        y -> [x,y]
                    )
                )
            ),
            pair -> jaro_winkler_similarity(pair[1], pair[2])
        )
    ) >= 0.5 as result
"""

duckdb.sql(sql)

```

</details>

Overall implementation looks good, we need to think a bit about the 'user facing' aspect of this:
- Best way to allow user to provide the distance function 
- How to deal with the distance vs similiaryt issue (for levenstein higher is worse, for others like jaro, higher is better)

For consistency we could consider:
[class JaroWinklerAtThresholds(ComparisonCreator):](https://github.com/jonnyshiuw/splink/blob/a4ff0fc67f46900eb631b31fcbff0074b2390e1e/splink/internals/comparison_library.py#L312-L318)
[class LevenshteinAtThresholds(ComparisonCreator):](https://github.com/jonnyshiuw/splink/blob/a4ff0fc67f46900eb631b31fcbff0074b2390e1e/splink/internals/comparison_library.py#L95-L101)

Where we use `distance_threshold_or_thresholds` vs `score_threshold_or_thresholds`

So I wonder whether we coud do something like:
[class AbsoluteDateDifferenceLevel(AbsoluteTimeDifferenceLevel):](https://github.com/jonnyshiuw/splink/blob/a4ff0fc67f46900eb631b31fcbff0074b2390e1e/splink/internals/comparison_level_library.py#L683)

where we have a base function that takes a `higher_is_more_similar` type argument, e.g. like [higher_](https://github.com/jonnyshiuw/splink/blob/a4ff0fc67f46900eb631b31fcbff0074b2390e1e/splink/internals/comparison_library.py#L372)

But then what's exposed to the user is actually two functions, called something like `ArrayStringDistanceLevel` and `ArrayStringSimilarityLevel`

Or we could even consider going as far as simply exposing `ArrayJaroWinklerLevel` `ArrayLevenshteinLevel` and maybe a couple more to the user

**Comment by RobinL** [comment]
Just to further explain the distance vs score, here's a slack convo i dug up where I asked the team!
image

**Comment by zmbc** [comment]
> But then what's exposed to the user is actually two functions, called something like ArrayStringDistanceLevel and ArrayStringSimilarityLevel which set the higher_is_more_similar argument for the user, so they don't have to think about it.

This sounds like the best compromise to me!

**Comment by zmbc** [comment]
@RobinL Circling back to how to design this, I actually would like to propose consistency with the [`DistanceFunctionAtThresholds`](https://github.com/moj-analytical-services/splink/blob/splink4_dev/splink/internals/comparison_library.py#L287C7-L294) class, rather than consistency with `JaroWinklerAtThresholds` etc.

Specifically, we'd have `PairwiseStringDistanceFunctionAtThresholds` and `PairwiseStringDistanceFunctionLevel`. These would each take a `distance_function` argument that would be validated to be one of the options "jaro," "levenshtein," etc and transpiled automatically. The value of this argument would then automatically set `higher_is_more_similar`.

This feels slightly weird given that "DistanceFunction" seems to imply that higher is more distant, i.e. less similar, but that is already an issue with the existing `DistanceFunctionAtThresholds`.

As an add-on (in a separate PR?) we could also have a more configurable `PairwiseDistanceFunctionAtThresholds` (and associated level) that did _not_ restrict the function used, did _not_ transpile the function name, and required the _user_ to specify `higher_is_more_similar`. I'd also like to give the user the ability to customize whether the min or max distance/similarity is used.

What do you think?

**Comment by RobinL** [comment]
Yeah - good spot re the existing issue/weirdness with `DistanceFunctionAtThresholds`

I think I agree with you - `PairwiseStringDistanceFunctionAtThresholds`  and `PairwiseStringDistanceFunctionLevel` seem like the best options and most consistent.  I also think it's a good idea to constrain the list of possibilities with a validated/enum, to keep a lid on complexity.  Also we can use `typing.Literal` so it auto-completes for the user
```
Literal["jaro", ...]
```


I think to minimise complexity I'm minded to leave it at that rather than go through the complexity of allowing any arbitrary sql function.  For highly customised things, the user can provide the comparison as a dict, and they can use `PairwiseStringDistanceFunctionLevel` to generate something close to what they need, that they can then manually edit.

**Comment by RobinL** [comment]
Superceded by https://github.com/moj-analytical-services/splink/pull/2517

**Pull request closed without merging**